### PR TITLE
change zookeeper package cause old one was deprecated

### PR DESCRIPTION
**Pull request by shawndwells**
Performs the same update of the Zookeeper package as https://github.com/Comcast/go-leaderelection/pull/6, except I've signed the contribution agreement. Credit to @sangheee

----

Repository https://github.com/samuel/go-zookeeper, which go-leaderelection relies on, is no longer maintained.
They recommend using https://github.com/go-zookeeper/zk instead.

So, changed the zookeeper package.

### CODE REVIEW

**Comment by shawndwells** [comment]
Hmm, the github bot says I haven't signed the CLA... but....

image

**Comment by shawndwells** [comment]
Looks like the CLA took a minute.

Paging @youngkin and @kristakhare for PR review.

**Comment by kristakhare** [comment]
Thank you! Pinging @youngkin for review. If they don't respond soon, I don't see any issues and can merge this for you.

**Comment by shawndwells** [comment]
Thanks @kristakhare. Will checkin later this week if we don't hear anything from @youngkin.

**Comment by youngkin** [comment]
Just to confirm, have all the tests passed? Otherwise it looks like this new library is just a drop-in replacement. If so, the changes look good to me.

I also no longer work for Comcast. It's probably more appropriate for  @kristakhare to merge this.

**Comment by lazandrei19** [comment]
Hi. Is there any update on this?

**Comment by shawndwells** [comment]
@youngkin correctly asked if the tests were ran/passed..... a comment which I did read, then went on vacation and forgot about until now 😬 

@lazandrei19 any chance you could run the test suite and update the PR here?
(for observers, @lazandrei19 and I work together)

**Comment by kristakhare** [comment]
Great, thanks for the update! I'll wait until the tests are confirmed then I can merge for you. Thank you both!

**Comment by lazandrei19** [comment]
@kristakhare Running the tests from `Election_test.go` passes with flying colors.
I also tried running the tests in the `integrationtest` folder, but those seem to depend on `github.comcast.com/viper-cog/clog`, so no dice there

**Comment by youngkin** [comment]
It looks like the only file that relies on `clog` is `integrationtest/leadercrashtest/mock_worker.go`. This file was overlooked when making this project opensource. This test doesn't look relevant. I don't recall why it was added. My suggestion is to remove this file and see if anything else prevents the integration tests from running.

**Comment by lazandrei19** [comment]
Sadly, even after removing everything related to `integrationtest/leadercrashtest/mock_worker.go`, I still wasn't able to get the integration tests to pass either on the original version or Shawn's version. It always panics with this message:
```golang
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1089679]

goroutine 4 [running]:
testing.tRunner.func1.2({0x11a85e0, 0x1346dd0})
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1212 +0x218
panic({0x11a85e0, 0x1346dd0})
        /usr/local/Cellar/go/1.17.2/libexec/src/runtime/panic.go:1038 +0x215
os.(*Process).signal(0x11a7380, {0x121e148, 0x1318440})
        /usr/local/Cellar/go/1.17.2/libexec/src/os/exec_unix.go:64 +0x39
os.(*Process).Signal(...)
        /usr/local/Cellar/go/1.17.2/libexec/src/os/exec.go:138
os.(*Process).kill(...)
        /usr/local/Cellar/go/1.17.2/libexec/src/os/exec_posix.go:68
os.(*Process).Kill(...)
        /usr/local/Cellar/go/1.17.2/libexec/src/os/exec.go:123
github.com/Comcast/goint.(*Kill).Go(0xc00007f140, 0xc000690150)
        /Users/alazarescu/code/golang/src/github.com/Comcast/goint/goint.go:511 +0x15b
github.com/Comcast/goint.(*Comp).Go(0xc000012a40, 0xc0000a2000)
        /Users/alazarescu/code/golang/src/github.com/Comcast/goint/goint.go:402 +0x223
github.com/Comcast/goint.(*Comp).Go(0xc000012e80, 0xc000016460)
        /Users/alazarescu/code/golang/src/github.com/Comcast/goint/goint.go:402 +0x223
github.com/Comcast/goint.(*Comp).Go(0xc000115f20, 0xc000016430)
        /Users/alazarescu/code/golang/src/github.com/Comcast/goint/goint.go:402 +0x223
github.com/Comcast/goint.(*Comp).Start(0xc000115f20)
        /Users/alazarescu/code/golang/src/github.com/Comcast/goint/goint.go:437 +0x45
github.com/Comcast/go-leaderelection/integrationtest.TestMain(0xc000120ea0)
        /Users/alazarescu/code/personal/src/github.com/Comcast/go-leaderelection/integrationtest/zktest_test.go:965 +0x452
testing.tRunner(0xc000120ea0, 0x11ed990)
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/Cellar/go/1.17.2/libexec/src/testing/testing.go:1306 +0x35a
```

**Comment by youngkin** [comment]
> Sadly, even after removing everything related to integrationtest/leadercrashtest/mock_worker.go, I still wasn't able t0 get the integrat`ion tests to pass either on the original version or Shawn's version. It always panics with this message: 

Well, that's disappointing. I'm going to take  a look. From the stack trace I'm guessing the problem is in [goint](https://github.com/Comcast/goint).

I don't have this or any of the dependencies on my computer. This may take a while. You're welcome to look into it as well. Let me know if do.

**Comment by youngkin** [comment]
@lazandrei19 What version of ZK are you using?

**Comment by youngkin** [comment]
@lazandrei19 would you please send me the GitHub link to your go-leaderelection fork? I need to work with this one vs. the Comcast one

**Comment by lazandrei19** [comment]
It looks like I'm running zookeeper server version 3.7.0

**Comment by youngkin** [comment]
Update: @lazandrei19 I have the tests running except one. I didn't have the problem you had. I did add 'go mod' support, that might have made a difference. I'll update again tomorrow.

**Comment by lazandrei19** [comment]
@youngkin considering that this PR will be (hopefully) merged into the `github.com/Comcast/go-leaderelection` repo shouldn't the top line of the `go.mod` be `module github.com/Comcast/go-leaderelection`?

**Comment by youngkin** [comment]
@lazandrei19 I meant to fork Shawn's repo and submit a PR to you. I obviously goofed and cloned it directly (don't know what I was thinking) and committed to the ZK fork. For that commit I'm going to comment on the changes as well as a couple of outstanding items.

**Comment by shawndwells** [comment]
hey @youngkin and @lazandrei19 - bringing the conversation back to the upstream PR - looks like all the changes from @youngkin were incorporated and the tests pass.

@youngkin if you could confirm here in the PR, we can ping @kristakhare to let them know we're ready to merge! 🍻

File: go.mod
Comment (youngkin) [inline go.mod:12]: This 'replace' directive needs to be removed then 'go mod tidy' needs to be run to make the necessary updates to 'go.sum'.

**Review by youngkin — CHANGES_REQUESTED** [review]

File: go.mod
Comment (lazandrei19) [inline go.mod]: Also this probably needs to be replaced with `github.com/Comcast/go-leaderelection`

Comment (youngkin) [reply go.mod]: Good catch, definitely!

Comment (shawndwells) [inline go.mod]:
```suggestion
module github.com/Comcast/go-leaderelection
```

Comment (shawndwells) [reply go.mod]: Done.

**Comment by youngkin** [comment]
@sangheee @shawndwells @lazandrei19 thanks for the contribution.

@kristakhare since I spent some time working they integration issues it seemed to make sense for me to go ahead and merge this PR.

**Comment by shawndwells** [comment]
Nice! Thank you @youngkin and @kristakhare !

**Comment by youngkin** [comment]
While working on this PR the integration tests failed with the error/panic ` Attempt to use unknown class`. I was running with the `-race` flag. Removing this flag allowed the test to run successfully... for a while. I started getting the error even without `-race`. Googling around I found [this problem](https://github.com/golang/go/issues/49138)  had been encountered with the `-race` flag. The gem that got my tests reliably working again was to prefix `go test` with an environment variable as follows:

` env MallocNanoZone=0 go test`

As per the [the issue](https://github.com/golang/go/issues/49138) above it appears to be limited to MacOS Monterey. I'm running 12.0.1 so I don't know if other releases have the same problem.

**Pull request merged**
